File: android/app/src/main/java/vn/team/freechat/MessageActivity.java
package vn.team.freechat;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import com.tvd12.ezyfoxserver.client.request.EzyRequest;

import tvd12.com.ezyfoxserver.client.R;
import vn.team.freechat.adapter.MessageListAdapter;
import vn.team.freechat.adapter.MessageListAdapters;
import vn.team.freechat.data.Message;
import vn.team.freechat.data.MessageReceived;
import vn.team.freechat.data.MessageSent;
import vn.team.freechat.mvc.Controller;
import vn.team.freechat.mvc.IController;
import vn.team.freechat.mvc.IView;
import vn.team.freechat.mvc.Mvc;
import vn.team.freechat.request.SendSystemMessageRequest;
import vn.team.freechat.request.SendUserMessageRequest;

/**
 * Created by tavandung12 on 10/5/18.
 */

public class MessageActivity extends AppActivity {

    private String targetContact;
    private IController messageController;
    private IController connectionController;

    private View loadingView;
    private View backButtonView;
    private TextView targetNameView;
    private TextView targetLastMessageView;
    private EditText messageInputView;
    private ImageButton sendButtonView;
    private RecyclerView messageListView;
    private MessageListAdapter messageListAdapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_message);
        initComponents();
        initViews();
        setViewsData();
        setupViewControllers();
    }

    @Override
    protected void onStart() {
        super.onStart();
        connectionController.addView("show-loading", new IView() {
            @Override
            public void update(Object data) {
                loadingView.setVisibility(View.VISIBLE);
            }
        });
        connectionController.addView("hide-loading", new IView() {
            @Override
            public void update(Object data) {
                loadingView.setVisibility(View.GONE);
            }
        });
        messageController.addView("add-message", new IView() {
            @Override
            public void update(Object data) {
                addMessage((MessageReceived)data);
            }
        });
    }

    @Override
    protected void onStop() {
        super.onStop();
        loadingView.setVisibility(View.GONE);
        messageController.removeView("add-message");
    }

    private void initComponents() {
        Mvc mvc = Mvc.getInstance();
        targetContact = getIntent().getStringExtra("targetContact");
        messageController = mvc.getController("message");
        connectionController = mvc.getController("connection");
    }

    private void initViews() {
        loadingView = findViewById(R.id.loading);
        View headerView = findViewById(R.id.header);
        backButtonView = headerView.findViewById(R.id.back);
        View targetBoxView = headerView.findViewById(R.id.targetBox);
        targetNameView = targetBoxView.findViewById(R.id.targetName);
        targetLastMessageView = targetBoxView.findViewById(R.id.lastMessage);
        messageListView = findViewById(R.id.messageList);
        View chatboxView = findViewById(R.id.chatbox);
        messageInputView = chatboxView.findViewById(R.id.messageInput);
        sendButtonView = chatboxView.findViewById(R.id.sendButton);
        targetNameView.setText(targetContact);
        targetLastMessageView.setText("");

        MessageListAdapters adapters = MessageListAdapters.getInstance();
        messageListAdapter = adapters.getAdapter(this, targetContact);
        messageListView.setAdapter(messageListAdapter);
        LinearLayoutManager messageListLayoutManager = newMessageListLayoutManager();
        messageListView.setLayoutManager(messageListLayoutManager);
    }

    private void setViewsData() {
        targetNameView.setText(targetContact);
        targetLastMessageView.setText("");
    }

    private void setupViewControllers() {
        backButtonView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                backToContactView();
            }
        });
        sendButtonView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                EzyRequest request;
                String messageText = messageInputView.getText().toString();
                if(targetContact.equalsIgnoreCase("System"))
                    request = new SendSystemMessageRequest(messageText);
                else
                    request = new SendUserMessageRequest(targetContact, messageText);
                app.send(request);
                Message message = new MessageSent(messageText, targetContact);
                addMessageItem(message);
                messageInputView.setText("");
            }
        });
    }

    private void addMessage(MessageReceived message) {
        addMessageItem(message);
        targetLastMessageView.setText(message.getMessage());
    }

    private void addMessageItem(Message message) {
        messageListAdapter.addItem(message);
        messageListAdapter.notifyDataSetChanged();
        messageListView.smoothScrollToPosition(messageListAdapter.getItemCount());
    }

    private void backToContactView() {
        onBackPressed();
    }

    private LinearLayoutManager newMessageListLayoutManager() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setStackFromEnd(true);
        return layoutManager;
    }
}

File: kotlin/app/src/main/java/vn/team/freechat_kotlin/MessageActivity.kt
package vn.team.freechat_kotlin

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.EditText
import android.widget.ImageButton
import android.widget.TextView
import com.tvd12.ezyfoxserver.client.request.EzyRequest
import vn.team.freechat_kotlin.adapter.MessageListAdapter
import vn.team.freechat_kotlin.adapter.MessageListAdapters
import vn.team.freechat_kotlin.data.Message
import vn.team.freechat_kotlin.data.MessageReceived
import vn.team.freechat_kotlin.data.MessageSent
import vn.team.freechat_kotlin.request.SendSystemMessageRequest
import vn.team.freechat_kotlin.request.SendUserMessageRequest

/**
 * Created by tavandung12 on 10/5/18.
 */

class MessageActivity : AppActivity() {

    private var targetContact: String? = null
    private var messageController: Controller?  = null
    private var connectionController: Controller? = null

    private var loadingView: View? = null
    private var backButtonView: View? = null
    private var headerView: View? = null
    private var targetBoxView: View? = null
    private var targetNameView: TextView? = null
    private var targetLastMessageView: TextView? = null
    private var chatboxView: View? = null
    private var messageInputView: EditText? = null
    private var sendButtonView: ImageButton? = null
    private var messageListView: RecyclerView? = null
    private var messageListAdapter: MessageListAdapter? = null
    private var messageListLayoutManager: LinearLayoutManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_message)
        initComponents()
        initViews()
        setViewsData()
        setupViewControllers()
    }

    override fun onStart() {
        super.onStart();
        connectionController?.addView("show-loading",  object : IView {
            override fun update(viewId: String, data: Any?) {
                loadingView?.visibility = View.VISIBLE
            }
        })
        connectionController?.addView("hide-loading", object : IView {
            override fun update(viewId: String, data: Any?) {
                loadingView?.visibility = View.GONE
            }
        })
        messageController?.addView("add-message", object : IView {
            override fun update(viewId: String, data: Any?) {
                val message = data as MessageReceived
                addMessageItem(message);
                targetLastMessageView?.text = message.message
            }
        })
    }

    override fun onStop() {
        super.onStop();
        loadingView?.visibility = View.GONE
        messageController?.removeView("add-message")
    }

    private fun initComponents() {
        val mvc = Mvc.getInstance()
        targetContact = intent.getStringExtra("targetContact")
        messageController = mvc.getController("message");
        connectionController = mvc.getController("connection")
    }

    private fun initViews() {
        loadingView = findViewById(R.id.loading)
        headerView = findViewById(R.id.header)
        backButtonView = headerView?.findViewById(R.id.back)
        targetBoxView = headerView?.findViewById(R.id.targetBox)
        targetNameView = targetBoxView?.findViewById(R.id.targetName)
        targetLastMessageView = targetBoxView?.findViewById(R.id.lastMessage)
        messageListView = findViewById(R.id.messageList)
        chatboxView = findViewById(R.id.chatbox)
        messageInputView = chatboxView?.findViewById(R.id.messageInput)
        sendButtonView = chatboxView?.findViewById(R.id.sendButton)
        targetNameView?.setText(targetContact)
        targetLastMessageView?.setText("")

        val adapters = MessageListAdapters.getInstance()
        messageListAdapter = adapters.getAdapter(this, targetContact!!)
        messageListView?.setAdapter(messageListAdapter)
        messageListLayoutManager = newMessageListLayoutManager()
        messageListView?.setLayoutManager(messageListLayoutManager)
    }

    private fun setViewsData() {
        targetNameView?.setText(targetContact)
        targetLastMessageView?.setText("")
    }

    private fun setupViewControllers() {
        backButtonView?.setOnClickListener({ _ ->
                backToContactView()
        });
        sendButtonView?.setOnClickListener({ _ ->
            val messageText = messageInputView?.getText().toString()
            val request : EzyRequest
            if(targetContact.equals("System", true))
                request = SendSystemMessageRequest(messageText)
            else
                request = SendUserMessageRequest(targetContact!!, messageText)
            app!!.send(request);
            val message = MessageSent(messageText, targetContact!!)
            addMessageItem(message);
            messageInputView?.setText("")
        });
    }

    private fun addMessageItem(message: Message) {
        messageListAdapter?.addItem(message);
        messageListAdapter?.notifyDataSetChanged()
        messageListView?.smoothScrollToPosition(messageListAdapter?.getItemCount()!!)
    }

    private fun backToContactView() {
        onBackPressed();
    }

    private fun newMessageListLayoutManager() : LinearLayoutManager {
        val layoutManager = LinearLayoutManager(this)
        layoutManager.stackFromEnd = true
        return layoutManager
    }
}